Add `force_readonly` option for owner of a `FOREIGN SERVER` in context of SQLite connection management

mkgrgis opened the pull request:
In this PR:

- Implementation of `force_readonly` option based on `sqlite3_open_v2` function `SQLITE_OPEN_READONLY` mode and can disallow any changes in SQLite database. Hence if `force_readonly` set to `true` any values of `updatable` options are ignored. If  `force_readonly` set to `false` or not set there is usual behaviour of `updatable` option like in `postgres_fdw`.
- Improve SQL query execution errors to separate context and hint parts
- Improve other global SQLite errors for connection to more common PostgreSQL style
- Fix some linter hints about trailing spaces in some files

t-kataym commented:
@mkgrgis , Please let me share my thought.  
Because the functionality for a user can be provided by `updatable` option, providing similar options may causes a confusion to user.
I don't understand the necessity of `force_readonly` option now, and cannot explain a decisive usage to distinguish them (what kind of situation user should use these option respectively) .
And I agree with @son-phamngoc 's [comment](https://github.com/pgspider/sqlite_fdw/pull/59#issuecomment-1586681304): it had better not to introduce a feature which assumes an uncertain case.

mkgrgis commented:
> what kind of situation user should use these option respectively

This option is usefully in multi-user environment for security reasons.
1. There is server with owner `u1` and `updatable` = `false`.
2. `u1` granted table rights to `u2`, `u3` etc.
3. Normally `u2` or `u3` can _overwrite_ `updatable` value for a table
4. If `force_readonly` = `true` only owner of the server can determine if change SQL are available for all represented objects from any SQLite DB file. If PostgreSQL works with important SQLite database this option helps to guarantee unchangeable content.

Please explain where have I mistaken, @t-kataym.

mkgrgis commented:
@t-kataym , let's continue review of this PR.

SQLite (and Firebird) data access model is not only SQL like in Oracle, MariaDB or PostgreSQL, because significantly based on OS _file_ permissions and I think we should give to foreign server owner some instruments with similar security effect without access to PostgreSQL server file system. In the new version of `README.md` there is full motivation in _Data modification access_ section. Could you please write me your opinion about low level access permissions in context of this PR?

t-kataym commented:
Hello, @mkgrgis 

I'm sorry for my late reply.

> This option is usefully in multi-user environment for security reasons.
> 1. There is server with owner u1 and updatable = false.
> 2. u1 granted table rights to u2, u3 etc.
> 3. Normally u2 or u3 can overwrite updatable value for a table
> 4. If force_readonly = true only owner of the server can determine if change SQL are available for all represented objects from any SQLite DB file. If PostgreSQL works with important SQLite database this option helps to guarantee unchangeable content.

I think you can control it with PostgreSQL feature without this PR.
If you wants not to allow u2 to update the table, you take away the privilege of UPDATE for the table from u2 on PostgreSQL core layer (not FDW layer) with REVOKE query.

Could you explain a situation which cannot realize on only PostgreSQL feature?

mkgrgis commented:
> Could you explain a situation which cannot realize on only PostgreSQL feature?

I am trying to prepare this SQL mutiuser use case.
I know, that for Firebird database file permissions and file open modes also are access control instruments and many years before there was many commits to Firebird to support readonly files as possible data source. SQLite by default supports readonly file mode, hence there is no programmatically or conceptual problem, we need only explaining of SQL context for this PR. I'll do this.

mkgrgis commented:
@t-kataym , I have made a research about most popular relational FDWs for PostgreSQL.
Such FDWs as `oracle_fdw`, `firebird_fdw`, `mysql_fdw`, `postgres_fdw` can have non-effective `updatable=true` for both server and table.
For example remote PostgreSQL user in `postgres_fdw` can have readonly data access property as result of remote `ALTER USER ... SET default_transaction_read_only = on;` or `SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY;`. Also in Oracle remote user, used for `oracle_fdw` can have no data modification permissions. In Firebird there is two popular methods to deny data modification: usual SQL `GRANT/REVOKE` commands and 
readonly `fdb` file on filesystem level similar to implementation for SQLite in this PR. 

In this PR I have implemented data access control _under remote connection level_. In case of SQLite there is only primitive file open mode, no SQL `GRANT/REVOKE` which can make non-effective `updatable=true` for both server and table in other relational DBs.
_I was wrong in https://github.com/pgspider/sqlite_fdw/pull/78#issuecomment-1611842054, this PR is not about multi-user context. For user grouping there is `GRANT role TO user`. This PR is about foreign object grouping as united entity with a data modification property value. In other FDWs this is result of data access grants of remote user, here we have only primitive file open mode._

son-phamngoc commented:
Hello @mkgrgis , thanks for your effort. I would like to share my opinion about this PR.

In my understanding, you are going to provide an option to allow user to control data access **at FDW level**, via 2 options `force_readonly` and `updatable`. GRANT/REVOKE can help us to control it, but at **object access level**.
At FDW level, we already have `updatable` option, but its behavior is not as the purpose of this PR.
For `updatable` option, the value of FOREIGN TABLE has higher priority than FOREIGN SERVER.
So, to control it, you defined `force_readonly` option, which ignores `updatable` when it is true. Is that correct?

In my opinion, I think it is confusing when providing 2 options to do the same thing.
From user viewpoint, it is confusing to configure 2 options which has similar meaning. There can be a lot of combination of those 2 options, for both FOREIGN TABLE and FOREIGN SERVER, and make user harder to determine correct behavior.
Actually, if we change the behavior of `updatable` to prioritize the value of FOREIGN SERVER, it will reach the purpose of this PR without creating any new option. However, it will be inconsistent compared to the behavior of postgres_fdw, which is our base FDW. Please consider this point.

mkgrgis commented:
Hello, @son-phamngoc ! Thanks for Your opinion, I'll comment it.

> In my understanding, you are going to provide an option to allow user to control data access **at FDW level**, via 2 options `force_readonly` and `updatable`. GRANT/REVOKE can help us to control it, but at **object access level**.

Please note, for other relational FDWs also there is remote `GRANT/REVOKE` in source databases for a remote user of a foreign server. This remote `GRANT/REVOKE` allow to group _all foreign server objects_ as potentially changeable or strongly read-only independent of `updatable` option values and dependant only on connection. This PR implements similar behaviour.

> So, to control it, you defined `force_readonly` option, which ignores `updatable` when it is true. Is that correct?

Yes. Grants or revokes of Oracle or PosrgreSQL remote user also ignores all of `updatable` options inside of FDW foreign objects.

In SQLite we have no "_remote_"-like global access control except to _file open mode_. The specifics of our FDW is that both access controls methods concentrates inside of FDW: universal but not obligatory `updatable` (based on connection, after connection) and something like access control before connection implemented as file open mode in this PR.

> In my opinion, I think it is confusing when providing 2 options to do the same thing. From user viewpoint, it is confusing to configure 2 options which has similar meaning.

Not the same thing, @son-phamngoc. With `force_readonly` I can **once** group all foreign server objects as read-only like `REVOKE INSERT, UPDATE, DELETE` for remote user in Oracle, MariaDB or Firebird. Without this option I can only generate a scripts by list of objects of a foreign server to make this objects read-only for a users or roles.

> There can be a lot of combination of those 2 options, for both FOREIGN TABLE and FOREIGN SERVER, and make user harder to determine correct behavior. 

Users of other relational FDWs can understand `force_readonly` option as _fully compatible with revoked change permissions for remote user_. I am sure, with other FDWs there is much of remote users without change rights and de facto ignorable `updatable` option values.

> Actually, if we change the behavior of `updatable` to prioritize the value of FOREIGN SERVER, it will reach the purpose of this PR without creating any new option. However, it will be inconsistent compared to the behavior of postgres_fdw, which is our base FDW.
> Please consider this point.

Yes, this was my first not correct proposal. To avoid inconsistency with `postgres_fdw` I propose to understand `force_readonly` option as some kind of _reduced remote access control, implemented inside of our FDW_ because SQLite works just with a file and have no code outside of FDW.
Now I can implement similar access control only on OS level - `postgres` OS user have no write access to SQLite file with a sensitive data and it's directory. But I thinks it's better to give similar possibility to `FOREIGN SERVER` owner.

son-phamngoc commented:
@mkgrgis Thanks for your explanation.

> With force_readonly I can once group all foreign server objects as read-only like REVOKE INSERT, UPDATE, DELETE for remote user in Oracle, MariaDB or Firebird. Without this option I can only generate a scripts by list of objects of a foreign server to make this objects read-only for a users or roles.

I understand that `force_readonly` can help you mark all objects which belong to a foreign server as read-only at once, instead of setting each object respectively.
However, introducing an option that ignores another option causes a confusion and makes complex. It is better to avoid it.
Therefore, we need a stronger reason to be able to accept this option.

Could you tell me an actual use case that you need to use `force_readonly` option?

mkgrgis commented:
> Could you tell me an actual use case that you need to use `force_readonly` option?

Yes, let's discuss real example.
There is collective PostgreSQL of college connected to SQLite file which is updated by outer system. Data schema of source SQLite is not very stable, but changeable approximate once in 3..4 weeks for 2..3 tables. Professor gives full data read access to students to all of SQLite tables (25..35), but sometimes edit source data anomalies in SQlite as `FOREIGN SERVER` owner. Now professor generates SQL like `REVOKE INSERT UPDATE DELETE` for all objects for student's role and works with PostgreSQL admin to allow temporary write access for `postgres` OS user to the SQLite file and the directory.
Also students have separate SQlite files for exam works connected to this collective PostgreSQL. Students can freely share and read all of the data, but before exam professor blocks (with help of OS admins) all personal student's SQLite files to avoid changes. With  `force_readonly` option professor can block changes in all of the student's SQLite files after he become owner of all of student's `FOREIGN SERVERS`. Ownership changed if a student tell his exam works in SQLite are ready.
This case is typical for remote relational databases, but with SQLite so called _remote connection control_ in not really remote, because SQLite file processing is not outside of PostgreSQL. I hope , @son-phamngoc, you have understood that so called _remote control access_ which allows so called _remote connection admins_ ignore `updatable` values is something independent of DB technical form - both needed for network or file DB.

son-phamngoc commented:
@mkgrgis  Could you help me to confirm which way of understanding matches your example?
1. There is only 1 FOREIGN SERVER, and the Professor is the owner of that FOREIGN SERVER, which can change value of `force_readonly`. The professor creates FOREIGN TABLE and GRANT INSERT UPDATE DELETE on FOREIGN TABLE for students.
The students have the permission to read and write FOREIGN TABLE. They do not have permission to access or change FOREIGN SERVER. 
For example: student 1 can read and write FOREIGN TABLE 1, student 2 can read and write FOREIGN TABLE 2.
In normal case, the professor does not set value of `force_readonly` (or set it to `false`) to allow students to access or update data.
Before exam, the professor change `force_readonly` to `true` to block all students from updating.
2. There are several FOREIGN SERVERs and the Professor is the owner of them.
The professor GRANT USAGE on FOREIGN SERVER to students, so that the students can access and modify server.
For example, FOREIGN SERVER 1 is granted to student 1, FOREIGN SERVER 2 is granted to student 2.
Students can create FOREIGN TABLE on their granted FOREIGN SERVER.
Before exam, firstly, the professor REVOKE USAGE on FOREIGN SERVER from students, so that the students cannot update the FOREIGN SERVER.
Next, he changes `force_readonly` of all FOREIGN SERVER to disallow student update data.

I think the 2nd matches your example.
If I have any misunderstanding, please explain to me.

mkgrgis commented:
@son-phamngoc , I am sorry for not clear clarification, really the example is near your **2** version.

> 1. There is only 1 FOREIGN SERVER, and the Professor is the owner of that FOREIGN SERVER, which can change value of `force_readonly`.

No, there are many foreign servers in PostgreSQL, but there is main `FOREIGN SERVER` which also works with outer system and have very big and _data quality sensitive_ SQLite file. This server owned by the Professor and usually is read only even for the Professor. Sometimes he changed some data.

>  The students have the permission to read and write FOREIGN TABLE. They do not have permission to access or change FOREIGN SERVER.

Students does not have permission to change any data of _main_ `FOREIGN SERVER` owned by the Professor. They can read and write any data on personal `FOREIGN SERVER`, sometimes student can write some data on `FOREIGN SERVER` of other student if there was `GRANT` from other student. By default there are no read restrictions on all of `FOREIGN SERVER`s.

> For example: student 1 can read and write FOREIGN TABLE 1, student 2 can read and write FOREIGN TABLE 2.

No. This limitation is very rare, even if a student tell to the Professor not to see some data of the main `FOREIGN SERVER`.

> 2\. There are several FOREIGN SERVERs and the Professor is the owner of them.

Yes. There is not only main `FOREIGN SERVER`, but also there are many `FOREIGN SERVER`s of students.

>         The professor GRANT USAGE on FOREIGN SERVER to students, so that the students can access and modify server.

Yes. Students works with data from the main `FOREIGN SERVER` and, sometimes, with data of `FOREIGN SERVER`s of other students.

>         For example, FOREIGN SERVER 1 is granted to student 1, FOREIGN SERVER 2 is granted to student 2.

Yes. This is group of `FOREIGN SERVER`s for students, usually with not big SQLite files.

>         Students can create FOREIGN TABLE on their granted FOREIGN SERVER.

Yes. But students cannot create or change a `FOREIGN TABLE` on the main `FOREIGN SERVER` owned by the Professor.

>         Before exam, firstly, the professor REVOKE USAGE on FOREIGN SERVER from students, so that the students cannot update the FOREIGN SERVER.
>         Next, he changes `force_readonly` of all FOREIGN SERVER to disallow student update data.

Yes. 
 
> I think the 2nd matches your example.

Yes, but this is not 1st and 2nd versions, this is only different aspects of the example. I am sorry for not good clarification, @son-phamngoc . I hope now we have no misunderstanding.

mkgrgis commented:
> In my opinion, I think it is confusing when providing 2 options ...

@son-phamngoc , in README added detailed description with diagnostics description in https://github.com/pgspider/sqlite_fdw/pull/78/commits/087b3a359462d7ad4a3c90b0889e5806f83cdd31 and https://github.com/pgspider/sqlite_fdw/pull/78/commits/6e6c74ee389cde0d6934ec2bde1c974367a12130

son-phamngoc commented:
Hello @mkgrgis , thanks for your answer. I think I understood the necessity of this option.

I would like to describe the necessity of this option like this:
```
Without `force_readonly` option, we cannot support A and B at the same time.
A. Grant user permission to create a foreign table
B. Revoke user permission to modify table data.
```
Could you add it to README?

About the implementation, I have not started review yet, but please let me share my thought.
I think we can simply reach the target of this MR by modifying `sqliteIsForeignRelUpdatable `.
When `sqliteIsForeignRelUpdatable` return False, which does not allow modification, the Postgres core will raise an error, and the modification stops immediately.
Therefore, I think the implementation of connection.c is redundant. There is no case that sqlite_fdw opens a read-only connection.
```
flags = flags | (entry->readonly ? SQLITE_OPEN_READONLY : SQLITE_OPEN_READWRITE);
rc = sqlite3_open_v2(dbpath, &entry->conn, flags, zVfs);
```
I remember that we have already discussed about that in https://github.com/pgspider/sqlite_fdw/pull/59.
If it is redundant, we should remove it.

mkgrgis commented:
> Without `force_readonly` option, we cannot support A and B at the same time.
> A. Grant user permission to create a foreign table
> B. Revoke user permission to modify table data.

Yes. There is no use case  difference with other FDWs, but there is difference of implementation. The description have completed in https://github.com/pgspider/sqlite_fdw/pull/78/commits/f92ee42aa26d178c28ac8e33c6ff0ad3c69056ce

> This option is useful if you need grant user permission to create a foreign tables on the foreign server and revoke user permission to modify any table data on this foreign server. This option with `true` value can disallow any write operations on foreign server table data through SQLite file readonly access mode. This option driven only by foreign server owner role can not be overwritten by any `updatable` option value. This is a strong restriction given by PostgreSQL foreign server owner user not to modify data in any foreign server tables. Also see [Connection to SQLite database file and access control](https://github.com/mkgrgis/sqlite_fdw/blob/draft_force_readonly_option/README.md#connection-to-sqlite-database-file-and-access-control) 

About the implementation...

> I think we can simply reach the target of this MR by modifying `sqliteIsForeignRelUpdatable `.

Yes, this is implemented, but not only. In other FDWs similar _"grant user permission to create a foreign table and revoke user permission to modify table data"_ conditions implementation based on remote database internal structures. I think in our case we also can use _remote database structures_ which is internal SQLite function in our case. Any `FOREIGN SERVER` option change calls a new connection and ends old connection for a new option combination. Also `force_readonly` is a classical security option and double implementation is useful for any security options. In our case `sqlite3_open_v2` with `SQLITE_OPEN_READONLY` needs additional 1-2 ms. If there will be found any CVE in the Postgres core or `sqlite3_open_v2` our double implementation will be not vulnerable. Maybe, I should add comments about this in the code?

mkgrgis commented:
> All the code merging into master branch should be tested, and make sure that it is effective. Please understand it.

Added tests allow us ensure there are no problems for any existed functions with `SQLITE_OPEN_READONLY`. Have I added destroyed or unusable TCs in this PR, @son-phamngoc ? Does I need to give you scripts with benchmark tests and my results with some millions `SQLITE_OPEN_READONLY` and  `SQLITE_OPEN_READWRITE` open connection operations?

> Seen from usability, the logic can be well implemented in `sqliteIsForeignRelUpdatable`. I do not see any logic problem.

Please allow me share opinion of antivirus or [SELinux](https://ja.wikipedia.org/wiki/Security-Enhanced_Linux) security administrator. There is security rule: "_you shouldn't collect unusable privileges_". PostgreSQL server is well-tested program, but it's hard to run it with `0` (root) user rights, because authors of PostgreSQL try to prevent unusable privileges for the server and running from `0` user documented as unwanted.
Let's imagine system monitoring log about opened files. A security administrator knows sensitive SQLite file as readonly for PostgreSQL server process, but there is `rw` (readwrite) alert in log in case of `force_readonly = 'true'`. This is false security alert and unusable write privilege. If `FOREIGN SERVER` options will be changed this will be a new connection with new privileges. This can allow security administrators to log when somebody changes `FOREIGN SERVER` rights. Also `SQLITE_OPEN_READONLY` mode doesn't destroy any `sqlite_fdw` work and doesn't decrease formal productivity of any SQL operations. What do you think about security administrator view and opinion, @son-phamngoc ?

son-phamngoc commented:
@mkgrgis Sorry for late respond.
We had an internal discussion. We understood your viewpoint and the reason for your modification.
We can accept this PR after finished reviewing and testing.

mkgrgis commented:
> @mkgrgis Sorry for late respond.

No problems, @son-phamngoc. I understand there is unusual for you external C programming in the PR. Really this is more to POSIX side, not to SQL.

> We had an internal discussion.

Does I need provide some external tests for file descriptors `r `and `rw` ? I don't understand how we can integrate this tests, because it based on shell commands out of PostgreSQL server context.

son-phamngoc commented:
> Does I need to provide some external tests for file access mode descriptors r and rw ? I don't understand how we can integrate this tests, because it based on shell commands out of PostgreSQL server context.

No, you don't need to provide more test.
I will start reviewing soon.

mkgrgis commented:
> I will start reviewing soon.

Thanks, @son-phamngoc , common description is updated https://github.com/pgspider/sqlite_fdw/pull/78#issue-1761727873

son-phamngoc reviewed (commented):
@mkgrgis I added some comments. Could you check them?

mkgrgis commented on sqlite_fdw--1.0.sql:
```diff
@@ -8,15 +8,24 @@ RETURNS fdw_handler
 
+COMMENT ON FOREIGN DATA WRAPPER sqlite_fdw
+IS 'SQLite foreign data wrapper';
+
```
I understand this inconsistency, but author of oracle_fdw doesn't think comments is critical element of SQL files. Changes here was reverted, please verify, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sqlite_fdw.h:
```diff
@@ -392,4 +392,4 @@ void sqlite_fdw_data_norm_functs_init(sqlite3* db);
 			binstr2int64(const char *s);
 
-#endif							/* SQLITE_FDW_H */
+#endif							/* SQLITE_FDW_H */
```
Reverted, please verify @son-phamngoc 
Also added in linter 
```
for f in ./*.h
do
	lastln=$(tail -c1 "$f" | hexdump -e '/1 "%02X"');
	echo "$lastln <- $f";
	[ "$lastln" != "0A" ] || echo "" >> $f;
done
```

son-phamngoc replied:
You added more an empty line to all files, so now all files have 2 empty lines.  
![image](https://github.com/pgspider/sqlite_fdw/assets/123929620/888735ff-6ff1-4e93-9a03-f70f6ea193c3)
Please revert it. All files have just 1 empty line at the end of file.

mkgrgis replied:
Maybe we have here misunderstanding, @son-phamngoc. Now I have
![изображение](https://github.com/pgspider/sqlite_fdw/assets/41448637/0bf73eb1-541c-471e-ae8d-0483ca9e9cb2)
in this case, no line 378.
`tail -c5 sqlite_data_norm.c | hexdump -e '2/1 "%02X"'` shows me `2F0A7D0A0A` where `7D`=`}`. What should it shows? No additional `0A`?

mkgrgis replied:
Maybe https://github.com/pgspider/sqlite_fdw/pull/78/commits/5ee9b0247d4cb5dd7e9306839557577f385908f3 gives normal ends of files?

son-phamngoc replied:
Thanks for your update. It's OK now. Please help me to resolve this comment.

mkgrgis commented on sqlite_fdw.c:
```diff
@@ -5691,14 +5694,28 @@ sqliteIsForeignRelUpdatable(Relation rel)
 
 		if (strcmp(def->defname, "updatable") == 0)
 			updatable = defGetBoolean(def);
+		if (strcmp(def->defname, "force_readonly") == 0)
```
Fixed, please verify, @son-phamngoc !

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sqlite_fdw.c:
```diff
@@ -5673,14 +5673,17 @@ static int
 
 	/*
 	 * By default, all sqlite_fdw foreign tables are assumed updatable.
+	 * If force_readonly option is set, table option 'updatable' is ignored
```
Updated, please recheck.

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on README.md:
```diff
@@ -170,13 +133,17 @@ SQLite `NULL` affinity always can be transparent converted for a nullable column
+- **force_readonly** as *boolean*, optional, default *false*
 
-- **truncatable** as *boolean*, optional, default *true*
+  This option is useful if you need grant user permission to create a foreign tables on the foreign server and revoke user permission to modify any table data on this foreign server. This option with `true` value can disallow any write operations on foreign server table data througth SQLite file readonly access mode. This option driven only by foreign server owner role can not be overwritten by a ...
```
Fixed, please verify, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on README.md:
```diff
@@ -170,13 +133,17 @@ SQLite `NULL` affinity always can be transparent converted for a nullable column
   Allows foreign tables to be truncated using the `TRUNCATE` command.
 
-- **keep_connections** as *boolean*, optional, default *true*
+- **keep_connections** as *boolean*, optional, default *false*
```
Maybe this was inserted from `README.md` for other FDW. Reverted, thanks, @son-phamngoc! This change have ho reason in `sqlite_fdw` context.

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on README.md:
```diff
@@ -233,6 +197,43 @@ in SQLite (mixed affinity case). Updated and inserted values will have this affi
   Indicates a column as a part of primary key or unique key of SQLite table.
 
+#### Datatypes
+**WARNING! The table above represents roadmap**, work still in progress. Untill it will be ended please refer real behaviour in non-obvious cases, where there is no ✔ or ∅ mark.
```
Fixed, please verify, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on README.md:
```diff
@@ -276,6 +278,34 @@ Following restrictions apply:
+Full list of OS-leveled conditions of data modification access to SQLite database file
+- Existed SQLite file is not corrupted by SQLite engine conditions.
+- All path elements of the file are readable (listable) for OS user of PostgreSQL server process.
+- The file and a direcotory of the file placed on readwrite filesystem. For example `sqashfs` is always read-only, remote `sshfs` can be read-only, a disk partition can be mounted in read-only mode etc.
```
Fixed, please verify, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on README.md:
```diff
@@ -276,6 +278,34 @@ Following restrictions apply:
+
+Full list of PostgreSQL-leveled conditions of data modification access to SQLite database file
+- The `FOREIGN SERVER` of the SQLite file have no `force_readonly` = `true` option value.
+- You have `USAGE` `GRANT` for the `FOREIGN SERVER`.
```
Changed to other sentence, please recheck, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on README.md:
```diff
@@ -276,6 +278,34 @@ Following restrictions apply:
+- The `FOREIGN TABLE` of SQLite table have no `updatable` = `false` option value.
+- If the `FOREIGN TABLE` have no `updatable` option, ensure `FOREIGN SERVER` have no `updatable` = `false` option value.
+
+Generally for `sqlite_fdw` access managment `FOREIGN SERVER` owner can be like _remote access manager_ for other FDWs.
```
Fixed, please verify, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on README.md:
```diff
@@ -276,6 +278,34 @@ Following restrictions apply:
+- If the `FOREIGN TABLE` have no `updatable` option, ensure `FOREIGN SERVER` have no `updatable` = `false` option value.
+
+Generally for `sqlite_fdw` access managment `FOREIGN SERVER` owner can be like _remote access manager_ for other FDWs.
+_Remote access manager_ can block any data modififcations in remote database for _remote user_ of a FDW. In this case SQLite have no user or separate access conceptions, hence `FOREIGN SERVER` owner combines _remote access manager_ role with inrernal PostgreSQL roles such as `FOREIGN SERVER` access managment.
```
Fixed, please verify, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on connection.c:
```diff
@@ -155,11 +155,6 @@ sqlite_get_connection(ForeignServer *server, bool truncatable)
-	 * If cache entry doesn't have a connection, we have to establish a new
-	 * connection.  (If sqlite_open_db has an error, the cache entry will
-	 * remain in a valid empty state, ie conn == NULL.)
-	 */
```
Reverted, there was no reason to remove this comment. Thanks, @son-phamngoc !

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sql/14.9/sqlite_fdw.sql:
```diff
@@ -633,28 +633,45 @@ SELECT * FROM case_exp WHERE CASE c3 COLLATE "C" WHEN c6 THEN true ELSE c3 < 'ba
 DELETE FROM case_exp;
 
--- updatable option test (github pull 59)
+-- readonly/readwrite tests
```
Expanded to all versions, @son-phamngoc ! Please recheck a file for 16.0 as copy source.

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on expected/16.0/sqlite_fdw.out:
```diff
@@ -1817,7 +1835,268 @@ HINT:  Constant value data type: "tsquery" in column "b"
+ALTER FOREIGN TABLE RO_RW_test OPTIONS (DROP updatable);
+--Testcase 364:
+ALTER FOREIGN TABLE RO_RW_test OPTIONS (DROP force_readonly);
+ERROR:  option "force_readonly" not found
```
Changed, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on expected/16.0/sqlite_fdw.out:
```diff
@@ -1817,7 +1835,268 @@ HINT:  Constant value data type: "tsquery" in column "b"
+(0 rows)
+
+-- End of RO/RW test
+-- readonly/readwrite test github pull 59
```
Such comments are deleted in all places of the test. Please ensure all is ok, @son-phamngoc

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on expected/16.0/sqlite_fdw.out:
```diff
@@ -1817,7 +1835,268 @@ HINT:  Constant value data type: "tsquery" in column "b"
+ALTER FOREIGN TABLE numbers ALTER COLUMN b TYPE varchar(255);
+-- updatable option test (github pull 59)
 DROP FOREIGN TABLE RO_RW_test;
+ERROR:  foreign table "ro_rw_test" does not exist
```
Deleted, @son-phamngoc . Please recheck.

son-phamngoc replied:
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sqlite_fdw.c:
```diff
@@ -5691,14 +5694,28 @@ sqliteIsForeignRelUpdatable(Relation rel)
+		if (strcmp(def->defname, "force_readonly") == 0)
+			readonly_db_file = defGetBoolean(def);
 	}
 	foreach(lc, table->options)
```
> Redundant block of code.

Yes, I have understand this. Is the code after https://github.com/pgspider/sqlite_fdw/pull/78/commits/e8ba199c60593c190843bfd85431ef912372d7b2 normal?

son-phamngoc replied:
Thank you. It's OK now.. Please help me to resolve this comment.

son-phamngoc commented on README.md:
```diff
@@ -276,6 +278,34 @@ Following restrictions apply:
+
+#### Data change access
+
+Data modification access in `sqlite_fdw` drived by both operating system and PostgreSQL.
```
Please update the following sentence, too.  
> Generally for `sqlite_fdw` access management `FOREIGN SERVER` owner can be like _remote access manager_ for other FDWs.

mkgrgis replied:
Fixed in https://github.com/pgspider/sqlite_fdw/pull/78/commits/fbe92e2fdd9625cb481e77924ab5d5bb9f78e2b3, please verify, @son-phamngoc !

son-phamngoc replied:
Thanks for your update. It's OK now. Please help me to resolve this comment.

son-phamngoc commented on connection.c:
```diff
@@ -247,10 +211,34 @@ sqlite_make_new_connection(ConnCacheEntry *entry, ForeignServer *server)
-	/* Try to make the connection */
-	entry->conn = sqlite_open_db(dbpath);
+	flags = flags | (entry->readonly ? SQLITE_OPEN_READONLY : SQLITE_OPEN_READWRITE);
+	rc = sqlite3_open_v2(dbpath, &entry->conn, flags, zVfs);
```
Thank you. I confirmed. Please help me to resolve this comment.

mkgrgis commented:
Thanks for 1st review, @son-phamngoc ! Let's begin a 2nd review round?

son-phamngoc commented:
Thanks for your fixing @mkgrgis.  
I see no more problem. @t-kataym will decide next step.

mkgrgis commented:
Ping, @t-kataym ?

t-kataym commented:
@mkgrgis Thank you for your contribution. I will accept the PR.
